### ISSUE TITLE
ENH: `numpy.isdtype` with custom dtypes

### ISSUE THREAD

**Issue by keewis** [numpy/numpy#27545]
### Proposed new feature or change:

The implementation of `numpy.isdtype` currently checks against an explicit list of kind names, and another one of dtypes (`numpy._core._type_aliases.allTypes`), and raises for all other objects. While the description of the exception ("dtype argument must be a NumPy dtype") makes sense for  for dtypes created for other array API implementing types, I don't think this is  the case for custom `numpy` dtypes which are explicitly written for `numpy`.

Incidentally, checking the new `StringDType` also raises on `numpy=2.1.1` (and as far as I can tell would even fail on [current `main`](https://github.com/numpy/numpy/blob/6f63230b54bc9910c849e78922d015e607b6848f/numpy/_core/numerictypes.py#L368-L468)):
```python
In [1]: import numpy as np
   ...: 
   ...: np.isdtype(np.dtypes.StringDType(), "real floating")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 3
      1 import numpy as np
----> 3 np.isdtype(np.dtypes.StringDType(), "real floating")

File .../numpy/_core/numerictypes.py:425, in isdtype(dtype, kind)
    423     dtype = _preprocess_dtype(dtype)
    424 except _PreprocessDTypeError:
--> 425     raise TypeError(
    426         "dtype argument must be a NumPy dtype, "
    427         f"but it is a {type(dtype)}."
    428     ) from None
    430 input_kinds = kind if isinstance(kind, tuple) else (kind,)
    432 processed_kinds = set()

TypeError: dtype argument must be a NumPy dtype, but it is a <class 'numpy.dtypes.StringDType'>.
```

This came up in pydata/xarray#9568, where someone wanted to use `xarray` with the dtypes defined by `ml_types`, which eventually failed because of the `numpy.isdtype` call.

So my question is: should `numpy.isdtype` support checking custom `numpy` dtypes? And if so, should it always return `False`, or do we need some kind of mechanism to include custom dtypes in the categories defined by the kind names?

**Comment by seberg** [comment]
> So my question is: should numpy.isdtype support checking custom numpy dtypes?

Yes this should definitely work (or at least not fail the way it does).  There is the caveat that making sure that `bfloat16` passes the `"real floating"` check may need a bit of work/fixes.

**Comment by Vednawale** [comment]
Proposed Change
Support for Custom Dtypes:

Enhancement: Allow numpy.isdtype to recognize and handle custom NumPy dtypes. This would make the function more flexible and user-friendly, especially for developers creating libraries that rely on custom dtypes.
Return Value: The function could return False for custom dtypes by default, unless explicitly defined otherwise. This would maintain backward compatibility while allowing for future extensibility.
Mechanism for Inclusion:

Custom Registration: Introduce a mechanism where users can register their custom dtypes with NumPy, allowing numpy.isdtype to recognize them. This could be done through a decorator or a registration function.
Dynamic Checking: Modify the internal logic of numpy.isdtype to check if the dtype is an instance of a registered custom dtype class.
Considerations
Backward Compatibility: Any changes should ensure that existing codebases relying on the current behavior of numpy.isdtype do not break. This could be achieved by maintaining the current checks while adding support for custom dtypes as an optional feature.
Documentation: Clear documentation should accompany any changes, explaining how users can create and register custom dtypes, and how numpy.isdtype will behave with these dtypes.
Community Feedback: Engaging with the NumPy community through discussions or proposals on platforms like GitHub can provide valuable insights and consensus on the best approach to implement this feature.
Conclusion
In summary, enhancing numpy.isdtype to support custom NumPy dtypes could significantly improve its usability and flexibility. The proposed changes would involve:

Allowing numpy.isdtype to recognize custom dtypes.
Returning False for unrecognized custom dtypes by default.
Implementing a registration mechanism for custom dtypes.

**Comment by mtsokol** [comment]
> > So my question is: should numpy.isdtype support checking custom numpy dtypes?
> 
> Yes this should definitely work (or at least not fail the way it does). There is the caveat that making sure that `bfloat16` passes the `"real floating"` check may need a bit of work/fixes.

I think that then we would need a `kind` attibute for dtypes to know to which kind the dtype belongs to (and for user-defined dtypes the user also needs to provide a `kind`, following [Array API taxonomy](https://data-apis.org/array-api/latest/API_specification/generated/array_api.isdtype.html)).

**Comment by ilan-gold** [comment]
+1 Running into the same issue, this time with moving to use the array api + `StringDtype`.  I think this is definitely a bug given the error message.

**Comment by ngoldbaum** [comment]
Hi all, sorry for letting this sit for so long. I completely missed it when it was filed in 2024. I'm working on a fix.
